File: 5、灰色关联分析/code1.py
import numpy as np

# 请输入初始矩阵[[120, 20, 5.5, 85], [130, 25, 5.7, 88], [128, 23, 5.6, 87], [140, 28, 5.9, 90], [138, 27, 6, 89], [150, 30, 6.2, 91]]
A = np.array(eval(input('请输入初始矩阵=')))  # 假设输入的形式是合法的Python列表形式

# 求出每一列的均值以供后续的数据预处理
Mean = np.mean(A, axis=0)

# 预处理后的矩阵
A_norm = A / Mean

print('预处理后的矩阵为：')
print(A_norm)

# 母序列
Y = A_norm[:, 0]

# 子序列
X = A_norm[:, 1:]

# 计算|X0-Xi|矩阵(在这里我们把X0定义为了Y)
absX0_Xi = np.abs(X - np.tile(Y.reshape(-1, 1), (1, X.shape[1])))

# 计算两级最小差a
a = np.min(absX0_Xi)

# 计算两级最大差b
b = np.max(absX0_Xi)

# 分辨系数取0.5
rho = 0.5

# 计算子序列中各个指标与母序列的关联系数
gamma = (a + rho * b) / (absX0_Xi + rho * b)

print('子序列中各个指标的灰色关联度分别为：')
print(np.mean(gamma, axis=0))

File: 5、灰色关联分析/code1.m
%% 灰色关联分析用于系统分析例题的讲解
clear;clc
% A=[120 20 5.5 85;130 25 5.7 88;128 23 5.6 87;140 28 5.9 90;138 27 6 89;150 30 6.2 91]
A=input('请输入初始矩阵=')  % 输入初始矩阵
Mean = mean(A);  % 求出每一列的均值以供后续的数据预处理
A = A ./ repmat(Mean,size(A,1),1);  %size(A,1)=6, repmat(Mean,6,1)可以将矩阵进行复制，复制为和A同等大小，然后使用点除（对应元素相除），这些在第一讲层次分析法都讲过
disp('预处理后的矩阵为：'); disp(A)
Y = A(:,1);  % 母序列
X = A(:,2:end); % 子序列
absX0_Xi = abs(X - repmat(Y,1,size(X,2)))  % 计算|X0-Xi|矩阵(在这里我们把X0定义为了Y)
a = min(min(absX0_Xi))    % 计算两级最小差a
b = max(max(absX0_Xi))  % 计算两级最大差b
rho = 0.5; % 分辨系数取0.5
gamma = (a+rho*b) ./ (absX0_Xi  + rho*b)  % 计算子序列中各个指标与母序列的关联系数
disp('子序列中各个指标的灰色关联度分别为：')
disp(mean(gamma))
